pkg/util/containers/v2/metrics/collector_docker_linux.go

```diff
@@ -0,0 +1,112 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+//go:build docker && linux
+// +build docker,linux
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/docker/docker/api/types"
+)
+
+func convertContainerStats(stats *types.Stats) *ContainerStats {
+	return &ContainerStats{
+		Timestamp: time.Now(),
+		CPU:       convertCPUStats(&stats.CPUStats),
+		Memory:    convertMemoryStats(&stats.MemoryStats),
+		IO:        convertIOStats(&stats.BlkioStats),
+		PID:       convertPIDStats(&stats.PidsStats),
+	}
+}
+
+func convertCPUStats(cpuStats *types.CPUStats) *ContainerCPUStats {
+	return &ContainerCPUStats{
+		Total:            util.Float64Ptr(float64(cpuStats.CPUUsage.TotalUsage)),
+		System:           util.Float64Ptr(float64(cpuStats.CPUUsage.UsageInKernelmode)),
+		User:             util.Float64Ptr(float64(cpuStats.CPUUsage.UsageInUsermode)),
+		ThrottledPeriods: util.Float64Ptr(float64(cpuStats.ThrottlingData.ThrottledPeriods)),
+		ThrottledTime:    util.Float64Ptr(float64(cpuStats.ThrottlingData.ThrottledTime)),
+	}
+}
+
+func convertMemoryStats(memStats *types.MemoryStats) *ContainerMemStats {
+	containerMemStats := &ContainerMemStats{
+		UsageTotal: util.Float64Ptr(float64(memStats.Usage)),
+		Limit:      util.Float64Ptr(float64(memStats.Limit)),
+	}
+
+	if rss, found := memStats.Stats["rss"]; found {
```

Comment:
IIRC, the `Stats` contain almost all cgroups memory data so you may be able to fill more fields?

Reply:
Here is the full content of the `Stats` map:
```go
map[string]uint64{
"active_anon":0x14a000,
"active_file":0x0,
"cache":0x0,
"dirty":0x0,
"hierarchical_memory_limit":0xaa12b000,
"hierarchical_memsw_limit":0x0,
"inactive_anon":0x0,
"inactive_file":0x0,
"mapped_file":0x0,
"pgfault":0x56a,
"pgmajfault":0x0,
"pgpgin":0x37b,
"pgpgout":0x20b,
"rss":0x142000,
"rss_huge":0x0,
"total_active_anon":0x14a000,
"total_active_file":0x0,
"total_cache":0x0,
"total_dirty":0x0,
"total_inactive_anon":0x0,
"total_inactive_file":0x0,
"total_mapped_file":0x0,
"total_pgfault":0x56a,
"total_pgmajfault":0x0,
"total_pgpgin":0x37b,
"total_pgpgout":0x20b,
"total_rss":0x142000,
"total_rss_huge":0x0,
"total_unevictable":0x0,
"total_writeback":0x0,
"unevictable":0x0,
"writeback":0x0}
```

Here are the fields of the `ContainerMemStats` structure:
```go
type ContainerMemStats struct {
        // Common fields
        UsageTotal   *float64
        KernelMemory *float64
        Limit        *float64
        Softlimit    *float64

        // Linux-only fields
        RSS       *float64
        Cache     *float64
        Swap      *float64
        OOMEvents *float64 // Number of events where memory allocation failed

        // Windows-only fields
        PrivateWorkingSet *float64
        CommitBytes       *float64
        CommitPeakBytes   *float64
}
```

The non-Windows fields that are not set yet are `KernelMemory`, `SoftLimit`, `Swap` and `OOMEvents`.
Do you think those fields could be computed from some of the `Stats` values ?

Reply:
You may need to handle both `cgroupv1` and `cgroupv2` because the fields are different and Docker reports them "raw":
https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1416
https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1510

I forgot that we did not put everything in the `ContainerMemStats` (from cgroups detailed stats), so indeed not a lot to get.

KernelMemory is mapped here for cgroups v2 (not available in v1): https://github.com/DataDog/datadog-agent/blob/7c519f08e2bb1bf20e409ba0e41e26663057eee0/pkg/util/cgroups/cgroupv2_memory.go#L40
So you could reuse the logic.

Swap is not returned by Docker regardless of cgroup versions.

OOMEvents (==FailCnt) is returned as a structured field by Docker here: https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1486

From `docker run` documentation:
```
      --memory-reservation bytes       Memory soft limit
```

SoftLimit seems to correspond to the `MemoryReservation` field, but not entirely sure: https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L89

Reply:
Thank you very much Vincent for all those details !

* I’ve added the computation of `KernelMemory` based on `kernel_stack` and `slab`.
* I’ve filled `OOMEvents` with `Failcnt`.

About the memory soft limit, I tried to start a container with a “magic” value for the `--memory-reservation`, but I couldn’t find that figure back from the docker stats API:

```
$ docker run --name sleep --memory-reservation 12345678 --memory 23456789 -d ubuntu:20.04 sleep infinity

c621d20d0e11ea52250e8ed0f65fa177ecaf53646b65400c3a61764ba127b3f4


$ curl -s --unix /var/run/docker.sock 'http://unix/containers/c621d20d0e11ea52250e8ed0f65fa177ecaf53646b65400c3a61764ba127b3f4/stats?stream=false' | jq .memory_stats

{
  "usage": 327680,
  "max_usage": 4558848,
  "stats": {
    "active_anon": 0,
    "active_file": 0,
    "cache": 0,
    "dirty": 0,
    "hierarchical_memory_limit": 23453696,
    "hierarchical_memsw_limit": 46911488,
    "inactive_anon": 0,
    "inactive_file": 0,
    "mapped_file": 0,
    "pgfault": 1287,
    "pgmajfault": 0,
    "pgpgin": 924,
    "pgpgout": 936,
    "rss": 0,
    "rss_huge": 0,
    "total_active_anon": 0,
    "total_active_file": 0,
    "total_cache": 0,
    "total_dirty": 0,
    "total_inactive_anon": 0,
    "total_inactive_file": 0,
    "total_mapped_file": 0,
    "total_pgfault": 1287,
    "total_pgmajfault": 0,
    "total_pgpgin": 924,
    "total_pgpgout": 936,
    "total_rss": 0,
    "total_rss_huge": 0,
    "total_unevictable": 0,
    "total_writeback": 0,
    "unevictable": 0,
    "writeback": 0
  },
  "limit": 23453696
}
```

I’ve started my container with:
* `--memory 23456789`
* `--memory-reservation 12345678`

whereas I could find the `--memory` value rounded up to the closest multiple of 4 Ki in `.memory_stats.limit` and `.memory_stats.stats.hierarchical_memory_limit`, I cannot find any field with a value close to the value of `--memory-reservation`.
Do you think this could also come from a cgroup v1 limitation ?

Reply:
No it's not a limitation it's just that Docker does not report it because it comes from another file `memory.soft_limit_in_bytes` (and `memory.low` cgroupv2).

But I think the value is the one from `MemoryReservation` from **Container Spec**. However I'm not sure pulling this metric is worth triggering another API Call.